### PR TITLE
Fix bug introduced in 48c699c27abf62565360a771d4fb935f7ab8f95a:

### PR DESCRIPTION
**Pull request by aboodman**
All our sample apps directly mutate the data returned by get() (and less commonly scan()). This was breaking change detection completely.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -59,7 +59,9 @@
     "out/replicache.dev.wasm"
   ],
   "dependencies": {
+    "@types/lodash-es": "^4.17.4",
```
Comment (aboodman) [inline package.json:62]: Pulling this in for `cloneDeep`. This adds about 5kb of code to the binary :-(, but writing a deep clone -- even if restricted to json types -- has to concern itself with cycle detection. I feel like that's more your wheelhouse than mine.

File: src/json.test.ts [hunk 2]
```diff
@@ -54,3 +54,14 @@ test('JSON deep equal', () => {
 
   t({a: 1, b: 2}, {b: 2, a: 1});
 });
+
+test('JSON deepFreeze', async () => {
+  const obj: any = deepFreeze({obj:{}, arr:[]});
```
Comment (aboodman) [inline src/json.test.ts:59]: This causes ts warnings that I can't figure out how to silence.

```
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
```

... doesn't work.

**Review by aboodman — COMMENTED** [review]
`prolly.Map` is way over complex now that it's just a sorted map. There's no reason for the separate `_pending` but I wanted to fix it and add tests with minimal changes before attempting change.

File: package.json (hunk 1 above)
Comment (arv) [reply package.json:62]: That seems fine for now.

File: src/json.test.ts (hunk 2 above)
Comment (arv) [reply src/json.test.ts:59]: I'll look

File: src/json.ts [hunk 3]
```diff
@@ -9,13 +9,25 @@ export type JSONValue =
   | Array<JSONValue>
   | JSONObject;
 
+/** Like `JSONValue` but deeply immutable */
```
Comment (arv) [inline src/json.ts:12]: The docs can link these if you write `[[JSONValue]]`

File: src/prolly/map.ts [hunk 4]
```diff
@@ -1,104 +1,131 @@
...
-    switch (p) {
-      case deleteSentinel:
-        return undefined;
-      case undefined:
-        return this._baseGet(key);
-      default:
-        return p;
-    }
+  has(key: string): boolean {
+    const {found} = this._binarySearch(key);
```
Comment (arv) [inline src/prolly/map.ts:83]: inline

File: src/prolly/map.ts [hunk 5]
```diff
@@ -1,104 +1,131 @@
...
-      return undefined;
-    }
-    const {found, index} = this._base.binarySearch(key);
+  get(key: string): ImmutableJSONValue | undefined {
+    const {found, index} = this._binarySearch(key);
     if (!found) {
       return undefined;
     }
-    return this._base.entries[index][1];
+    const [, value] = this._entries[index];
```
Comment (arv) [inline src/prolly/map.ts:92]: inline

File: src/prolly/map.ts [hunk 6]
```diff
@@ -1,104 +1,131 @@
...
+    const {found, index} = this._binarySearch(key);
+    if (found) {
+      this._entries.splice(index, 1);
+    }
+    this._changedKeys.add(key);
   }
 
   entries(): IterableIterator<Entry> {
-    return new Iter(this._base, this._pending);
+    return this._entries[Symbol.iterator]();
```
Comment (arv) [inline src/prolly/map.ts:116]: or jusr `.values()`

File: src/prolly/map.ts [hunk 7]
```diff
@@ -1,104 +1,131 @@
...
     return this.entries();
   }
 
   async flush(write: dag.Write): Promise<string> {
-    const newBase = new Leaf([...this]);
-    await write.putChunk(newBase.chunk);
-    this._base = newBase;
-    this._pending.clear();
-    return this._base.chunk.hash;
+    // TODO: how come ts doesn't complain about passing the mutable array?
```
Comment (arv) [inline src/prolly/map.ts:124]: I played around with this yesterday and it makes sense but it is not very helpful for us.

File: src/prolly/map.ts [hunk 8]
```diff
@@ -1,104 +1,131 @@
...
+    const chunk = dag.Chunk.new(this._entries, []);
+    await write.putChunk(chunk);
+    this._entries = [...this._entries];
```
Comment (arv) [inline src/prolly/map.ts:127]: do we need a new array here?

**Comment by arv** [comment]
I like how this gets rid of the leaf iterator! But maybe that was not the intent of this PR

File: src/json.ts (hunk 3 above)
Comment (aboodman) [reply src/json.ts:12]: done

File: src/prolly/map.ts (hunk 4 above)
Comment (aboodman) [reply src/prolly/map.ts:83]: This isn't in the newer PR, but for my info, can you explain what you mean here? I don't understand your comment.

File: src/prolly/map.ts (hunk 7 above)
Comment (aboodman) [reply src/prolly/map.ts:124]: I filed https://github.com/rocicorp/replicache/issues/483.

File: src/json.test.ts (hunk 2 above)
Comment (aboodman) [reply src/json.test.ts:59]: I figured it out (actually just clicked the little hint from vs code, which figured it out for me).

**Comment by aboodman** [comment]
> I like how this gets rid of the leaf iterator! But maybe that was not the intent of this PR

Yeah I would like to do it, but this PR got out of hand, so I narrowed it and moved that idea to: https://github.com/rocicorp/replicache/issues/481

File: src/prolly/map.ts (hunk 4 above)
Comment (arv) [reply src/prolly/map.ts:83]: I just wanted it to be: `return this._binarySearch(key).found`

BTW. I'm applying this change.

**Pull request closed without merging**